### PR TITLE
Use a typed UUID for Dataset

### PR DESCRIPTION
**Pull request by jmpesp**

### CODE REVIEW

File: nexus/db-queries/src/db/datastore/dataset.rs [hunk 1]
```diff
@@ -27,7 +28,7 @@ use chrono::Utc;
 use diesel::prelude::*;
 use diesel::upsert::excluded;
 use futures::FutureExt;
-use nexus_db_model::DatasetKind;
+use nexus_db_model::DatasetKind as DbDatasetKind;
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/dataset.rs:31]: Is this rename necessary?

File: nexus/db-queries/src/db/datastore/dataset.rs [hunk 2]
```diff
@@ -230,8 +234,10 @@ impl DataStore {
             let batch = self
                 .dataset_list(opctx, filter_kind, &p.current_pagparams())
                 .await?;
-            paginator =
-                p.found_batch(&batch, &|d: &nexus_db_model::Dataset| d.id());
+            paginator = p
+                .found_batch(&batch, &|d: &nexus_db_model::Dataset| {
+                    *d.id().as_untyped_uuid()
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/dataset.rs:239]: IIRC there's also

```suggestion
                    d.id().into_untyped_uuid()
```

File: nexus/db-queries/src/db/datastore/mod.rs [hunk 3]
```diff
@@ -1043,7 +1044,8 @@ mod test {
 
             for (dataset, region) in dataset_and_regions {
                 // Must be 3 unique datasets
-                assert!(disk_datasets.insert(dataset.id()));
+                let dataset_id = dataset.id();
+                assert!(disk_datasets.insert(*dataset_id.as_untyped_uuid()));
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/mod.rs:1048]:
```suggestion
                assert!(disk_datasets.insert(dataset_id.into_untyped_uuid()));
```

File: nexus/test-utils/src/lib.rs [hunk 4]
```diff
@@ -254,8 +255,10 @@ impl RackInitRequestBuilder {
         let zone = self
             .internal_dns_config
             .host_zone(
-                // TODO-cleanup use TypedUuid everywhere
-                OmicronZoneUuid::from_untyped_uuid(dataset_id),
+                // XXX why is zone id being created from dataset id?
```
Comment (jgallagher) [inline nexus/test-utils/src/lib.rs:258]: This doesn't seem right; maybe related to https://github.com/oxidecomputer/omicron/pull/7000#discussion_r1830280079 (cc @smklein)?

**Review by jgallagher — COMMENTED** [review]
Thanks for this; LGTM except for that (pre-existing) confusion of zone ID / dataset ID. We probably want to fix that, I think?

File: clients/nexus-client/src/lib.rs [hunk 5]
```diff
@@ -51,6 +51,7 @@ progenitor::generate_api!(
         TypedUuidForDemoSagaKind = omicron_uuid_kinds::DemoSagaUuid,
         TypedUuidForDownstairsKind = omicron_uuid_kinds::TypedUuid<omicron_uuid_kinds::DownstairsKind>,
         TypedUuidForPropolisKind = omicron_uuid_kinds::TypedUuid<omicron_uuid_kinds::PropolisKind>,
+        TypedUuidForDatasetKind = omicron_uuid_kinds::TypedUuid<omicron_uuid_kinds::DatasetKind>,
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply clients/nexus-client/src/lib.rs]: right, done in 994cc4cd0

File: nexus/db-model/src/dataset.rs [hunk 6]
```diff
@@ -45,7 +47,7 @@ pub struct Dataset {
     ip: Option<ipv6::Ipv6Addr>,
     port: Option<SqlU16>,
 
-    pub kind: DatasetKind,
+    pub kind: DbDatasetKind,
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/dataset.rs]: cool, done in 0d24cc6f2

File: nexus/db-queries/src/db/datastore/dataset.rs (hunk 1 above)
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/dataset.rs]: no! 61b0be544

File: nexus/db-queries/src/db/datastore/dataset.rs (hunk 2 above)
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/dataset.rs]: d907fcbc4

File: nexus/db-queries/src/db/datastore/mod.rs (hunk 3 above)
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/mod.rs]: d907fcbc4

File: nexus/test-utils/src/lib.rs (hunk 4 above)
Comment (jmpesp) [reply nexus/test-utils/src/lib.rs]: 12da0ba8e

**Review by jgallagher — APPROVED** [review]
